client/hwlib/src/collectors/cpuinfo.rs

```diff
@@ -0,0 +1,176 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Written by:
+ *        Nadzeya Hutsko <nadzeya.hutsko@canonical.com>
+ */
+
+use anyhow::Result;
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::{self, BufRead};
+use std::str::FromStr;
+
+#[derive(Debug)]
+pub struct CpuInfo {
+    pub platform: String,
+    pub count: usize,
+    pub cpu_type: String,
+    pub model: String,
+    pub model_number: String,
+    pub model_version: String,
+    pub model_revision: String,
+    pub cache: i64,
+    pub bogomips: i64,
+    pub speed: u64,
+    pub other: String,
+}
+
+/// Parse /proc/cpuinfo the same way it's done in checkbox
+/// https://github.com/canonical/checkbox/blob/a8d5e9d/checkbox-support/checkbox_support/parsers/cpuinfo.py
+pub fn parse_cpuinfo(cpuinfo_filepath: &'static str) -> Result<CpuInfo> {
+    let file = File::open(cpuinfo_filepath)?;
+    let reader = io::BufReader::new(file);
+
+    let mut attributes: HashMap<String, String> = HashMap::new();
+    let mut count = 0;
+
+    for line in reader.lines() {
+        let line = line?;
+        if line.trim().is_empty() {
+            continue;
+        }
+
+        let parts: Vec<&str> = line.split(':').collect();
+        if parts.len() == 2 {
+            let key = parts[0].trim().to_string();
+            let value = parts[1].trim().to_string();
+
+            if key == "processor" {
+                count += 1;
+            }
+
+            attributes.insert(key, value);
+        }
+    }
+
+    let machine = std::env::consts::ARCH.to_string();
+    let speed = parse_speed(attributes.get("cpu MHz")).unwrap_or(0);
+
+    let model = attributes
+        .get("Model")
+        .or_else(|| attributes.get("model name"))
+        .unwrap_or(&machine)
+        .to_string();
+
+    let cpu_info = CpuInfo {
+        platform: machine.clone(),
+        count,
+        cpu_type: attributes.get("vendor_id").unwrap_or(&machine).to_string(),
+        model,
+        model_number: attributes
+            .get("cpu family")
+            .unwrap_or(&"".to_string())
+            .to_string(),
+        model_version: attributes
+            .get("model")
+            .unwrap_or(&"".to_string())
+            .to_string(),
+        model_revision: attributes
+            .get("stepping")
+            .unwrap_or(&"".to_string())
+            .to_string(),
+        cache: parse_cache_size(attributes.get("cache size"))?,
+        bogomips: parse_bogomips(attributes.get("bogomips"))?,
+        speed,
+        other: attributes
+            .get("flags")
+            .unwrap_or(&"".to_string())
+            .to_string(),
+    };
+
+    Ok(cpu_info)
+}
+
+/// Parse CPU frequency in MHz as it's done in checkbox
+/// https://github.com/canonical/checkbox/blob/a8d5e9/providers/resource/bin/cpuinfo_resource.py#L56-L63
```

Comment:
This doc isn't quite true, this function _reads_ and parses the CPU frequency. I think with the type changes I suggest below though, the first line is somewhat redundant. Is it part of the contract that this is done in the same way as checkbox, or is this an implementation detail? If so, I think it would be nicer to put this information just inside the function

Reply:
>  Is it part of the contract that this is done in the same way as checkbox, or is this an implementation detail?

We need to collect the information about the hardware the same way checkbox does it. Here are the diagrams that explain the workflow, I hope it helps to understand why:

```mermaid
sequenceDiagram
    Checkbox ->> DUT: Collect the information about device under test
    Checkbox ->> C3: Send the report about the DUT
    C3 ->> C3: Parse the report to DB
    hwapi server ->> C3: Collect the information about certified devices
    hwapi server ->> hwapi server: Put this information into local DB replica
```
```mermaid
sequenceDiagram
    hwapi client ->> User's machine: Collect the info about hardware
    hwapi client ->> hwapi server: Send the request to check certification status
    hwapi server ->> hwapi server: Check certification status using local DB replica
    hwapi server ->> hwapi client: Send certification status response
```

So, C3 DB is populated by the checkbox submissions, that's why the data needs to be collected the same way as checkbox does it

Reply:
I've updated the doc-string in ad45b88

Reply:
Ah gotcha, thanks for the explanation!